tree/ntuple/v7/src/RField.cxx

```diff
@@ -916,6 +923,181 @@ void ROOT::Experimental::RClassField::AcceptVisitor(Detail::RFieldVisitor &visit
+   : ROOT::Experimental::Detail::RFieldBase(fieldName, className, ENTupleStructure::kCollection, false /* isSimple */),
+     fNWritten(0)
+{
+   if (classp == nullptr)
+      throw RException(R__FAIL("RField: no I/O support for collection proxy type " + std::string(className)));
+
+   fProxy = classp->GetCollectionProxy();
+   if (!fProxy)
+      throw RException(R__FAIL(std::string(className) + " has no associated collection proxy"));
+   if (fProxy->HasPointers())
+      throw RException(R__FAIL("collection proxies whose value type is a pointer are not supported"));
+   if (fProxy->GetProperties() & TVirtualCollectionProxy::kIsAssociative)
+      throw RException(R__FAIL("associative collections not supported"));
+
+   std::unique_ptr<ROOT::Experimental::Detail::RFieldBase> itemField;
+   if (auto valueClass = fProxy->GetValueClass()) {
+      // Element type is a class
+      itemField = RFieldBase::Create("_0", valueClass->GetName()).Unwrap();
+   } else {
+      switch (fProxy->GetType()) {
+      case EDataType::kChar_t:   itemField = std::make_unique<RField<char>>("_0"); break;
+      case EDataType::kUChar_t:  itemField = std::make_unique<RField<std::uint8_t>>("_0"); break;
+      case EDataType::kShort_t:  itemField = std::make_unique<RField<std::int16_t>>("_0"); break;
+      case EDataType::kUShort_t: itemField = std::make_unique<RField<std::uint16_t>>("_0"); break;
+      case EDataType::kInt_t:    itemField = std::make_unique<RField<std::int32_t>>("_0"); break;
+      case EDataType::kUInt_t:   itemField = std::make_unique<RField<std::uint32_t>>("_0"); break;
+      case EDataType::kLong_t:
+      case EDataType::kLong64_t:
+         itemField = std::make_unique<RField<std::int64_t>>("_0");
+         break;
+      case EDataType::kULong_t:
+      case EDataType::kULong64_t:
+         itemField = std::make_unique<RField<std::uint64_t>>("_0");
+         break;
+      case EDataType::kFloat_t:  itemField = std::make_unique<RField<float>>("_0"); break;
+      case EDataType::kDouble_t: itemField = std::make_unique<RField<double>>("_0"); break;
+      case EDataType::kBool_t:   itemField = std::make_unique<RField<bool>>("_0"); break;
+      default:
+         throw RException(R__FAIL("unsupported value type"));
+      }
+   }
+   fItemSize = itemField->GetValueSize();
+   Attach(std::move(itemField));
+}
+
+std::unique_ptr<ROOT::Experimental::Detail::RFieldBase>
+ROOT::Experimental::RCollectionClassField::CloneImpl(std::string_view newName) const
+{
+   return std::unique_ptr<RCollectionClassField>(
+      new RCollectionClassField(newName, GetType(), fProxy->GetCollectionClass()));
+}
+
+std::size_t ROOT::Experimental::RCollectionClassField::AppendImpl(const Detail::RFieldValue &value)
+{
+   TVirtualCollectionProxy::TPushPop RAII(fProxy, value.GetRawPtr());
+   std::size_t nbytes = 0;
+   auto count = fProxy->Size();
+   for (unsigned i = 0; i < count; ++i) {
+      auto itemValue = fSubFields[0]->CaptureValue(fProxy->At(i));
```

Comment:
`TVirtualCollectionProxy::At` is inefficient in case of `list` (or anything not a vector).  This will work but as an upcoming PR we ought to replace those usage by the usage of iterators.  See the usage of `TVirtualCollectionIterators` in `TBranchElement` (or the lower level iterator interfaces in `TVirtualCollectionProxy`)

Reply:
Another (essential?) future optiomization is to specialize those routine for `std::vector` based on the fact that the content is in a contiguous memory location.

Reply:
> `TVirtualCollectionProxy::At` is inefficient in case of `list` (or anything not a vector). This will work but as an upcoming PR we ought to replace those usage by the usage of iterators. See the usage of `TVirtualCollectionIterators` in `TBranchElement` (or the lower level iterator interfaces in `TVirtualCollectionProxy`)

This is now https://github.com/root-project/root/issues/11671.